### PR TITLE
Redraw more frequently during auto-move

### PR DESCRIPTION
**Pull request by Qrox**
#### Summary
Bugfixes "Redraw more frequently during auto-move"

#### Purpose of change
Previously during auto-move, the window was redrawn every 5 minutes, making the display jumpy. This makes it redraw once per 10 seconds during auto-travel and every move during auto-move or loot sorting.

#### Describe the solution
Redraw once per 10 seconds when the player `has_distant_destination`, and every move during local auto-move. Also make auto-move and auto-drive interruptible by calling `handle_key_blocking_activity` instead of `pump_events`.

#### Testing
Auto-moving now feels much more smooth in-game.

### CODE REVIEW

**Comment by actual-nh** [comment]
@eltank: This might help with the jumpy videos.

**Comment by eltank** [comment]
> @eltank: This might help with the jumpy videos.

I am checking it out / playtesting.

**Comment by eltank** [comment]
Autowalk does look much smoother now. However, autodrive seems slower than I expect. I need to compare with `master`.

https://user-images.githubusercontent.com/8000047/129435187-293e5895-9144-4dc5-a2c2-b72b5cece561.mp4

https://user-images.githubusercontent.com/8000047/129435189-1b7cc3f9-5616-43dd-9b05-1bcdf7aed429.mp4

**Comment by actual-nh** [comment]
The Clang 12 (Tiles, ASan) build (but not the Basic one - possibly due to it not being on Tiles) failed because, during the vehicle ramp test, the code tried to call input_manager::get_input_event; this is the final one of the failures:
```
(~[slow] ~[.])=> ../tests/vehicle_ramp_test.cpp:279: FAILED:
(~[slow] ~[.])=>   {Unknown expression after the reported line}
(~[slow] ~[.])=> due to unexpected exception with message:
(~[slow] ~[.])=>   input_manager::get_input_event called in test mode
```
Perhaps a problem with making it (more) resizable?

**Comment by eltank** [comment]
When autodriving it takes about 50% longer to get to the destination compared to before this change.

**Comment by actual-nh** [comment]
> When autodriving it takes about 50% longer to get to the destination compared to before this change.

Possibly refresh every 60 (game) seconds during autodriving, as a compromise between 10 seconds and 5 minutes?

**Comment by eltank** [comment]
> > When autodriving it takes about 50% longer to get to the destination compared to before this change.
> 
> Possibly refresh every 60 (game) seconds during autodriving, as a compromise between 10 seconds and 5 minutes?

Autodrive is not doing any frame skipping at all (which seems ok). It's just that rendering is slower now for some reason. I'll upload a video comparison soon.

**Comment by eltank** [comment]
Baseline (built at 837b2a89d69ed491a708b912e4af209d78b959f5): 61s [real time]
Test (built at 837b2a89d69ed491a708b912e4af209d78b959f5 plus this PR patched in): 100s [real time]

https://user-images.githubusercontent.com/8000047/129435826-0dd3fc0c-c50b-4fc2-a401-b71ae3c7b5db.mp4

https://user-images.githubusercontent.com/8000047/129435831-5e31e7d4-4ba5-4892-a5cb-140c0ebfc545.mp4

I love the improvements, but please don't make autodrive slower.

**Comment by kevingranade** [comment]
This was specifically a hack to handle excessively low frame rates, to un-hack it, it needs some serious performance work.

**Comment by PatrikLundell** [comment]
Does this affect other activities as well, such as when e.g. butchering or pulling up a gate, get a message something is seen and if you want to quit, look around, see nothing nearby (still too far away to detect you), continue, and then get told you're suddenly attacked?

**Comment by Qrox** [comment]
@actual-nh It now calls `input_manager::get_input_event` to handle player interrupt during auto-move. Testing `test_mode` in `handle_key_blocking_activity` should fix it.

@eltank I'll look into the performance issue.

@kevingranade I'm aware of that, but I think the frequency can be increased a bit to avoid the confusing jumps. If it turns out to affect frame rate too much I can decrease the redraw frequency.

@PatrikLundell During activities other than auto-move it is still only redrawn every 5 minutes, so you'll still likely to get suddenly attacked. Though it now redraws the terrain once before showing the cancel query, so if the monster is already close it will display properly.

**Comment by PatrikLundell** [comment]
Thanks Qrox, that's a significant improvement, as you can now actually see the situation as it is when the cancel query pops up, rather than what the situation was 5 minutes ago.

**Comment by Qrox** [comment]
Well, I can't seem to reproduce the performance drop. In both release and debug builds the time to auto-drive was roughly the same. In one run it took double time but that happened without applying this change so I think it was unrelated.

@eltank Does it take more time every time you auto-drive with this change? If so could you post your system info and a save that reproduces the performance drop?

**Comment by eltank** [comment]
> Well, I can't seem to reproduce the performance drop. In both release and debug builds the time to auto-drive was roughly the same. In one run it took double time but that happened without applying this change so I think it was unrelated.
> 
> @eltank Does it take more time every time you auto-drive with this change? If so could you post your system info and a save that reproduces the performance drop?

Thanks for looking into it. The slowdown appears to be a constant factor. I suspect there's an extra call to the renderer on each move compared to before (i.e. it increases the frames-per-turn). If you have a fast computer you might not notice it.

[DriveTest.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/6987246/DriveTest.zip)

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: 6ecd3158 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]

**Comment by eltank** [comment]
I wrote this hack to count the number of draw() calls per turn, but it was 3 with and without your change. So I guess that's not it.

```diff
diff --git a/src/cata_tiles.cpp b/src/cata_tiles.cpp
index f40ed46..4ccdb05 100644
--- a/src/cata_tiles.cpp
+++ b/src/cata_tiles.cpp
@@ -1087,6 +1087,15 @@
     if( !g ) {
         return;
     }
+    static int num_calls = 1;
+    static time_point turn = calendar::turn;
+    if( calendar::turn == turn ) {
+        num_calls++;
+    } else {
+        std::cout << "Num draw() calls last turn: " << num_calls << std::endl;
+        num_calls = 1;
+        turn = calendar::turn;
+    }
 
 #if defined(__ANDROID__)
     // Attempted bugfix for Google Play crash - prevent divide-by-zero if no tile
```

**Comment by eltank** [comment]
Ok, I found the source of the performance regression. If I revert your change to map.cpp (adding a call to `handle_key_blocking_activity` in `vehicle::move()`) then performance goes back to normal.

**Comment by eltank** [comment]
It seems that what's really slowing down the game is an increase in the number of `refresh_display` calls per turn, from ~3-4 to 7.

```diff
diff --git a/src/sdltiles.cpp b/src/sdltiles.cpp
index d7a68d6..156b6c0 100644
--- a/src/sdltiles.cpp
+++ b/src/sdltiles.cpp
@@ -541,6 +541,15 @@
     if( test_mode ) {
         return;
     }
+    static int num_calls = 1;
+    static time_point turn = calendar::turn;
+    if( calendar::turn == turn ) {
+        num_calls++;
+    } else {
+        std::cout << "Num refresh_display() calls last turn: " << num_calls << std::endl;
+        num_calls = 1;
+        turn = calendar::turn;
+    }
 
     // Select default target (the window), copy rendered buffer
     // there, present it, select the buffer as target again.
```

**Comment by eltank** [comment]
If I print the turn numbers, the pattern is even stranger:

Before your change:
```
Num draw() calls last turn: 3  turn=5213201
Num draw() calls last turn: 3  turn=5213202
Num refresh_display() calls last turn: 4  turn=5213202
Num refresh_display() calls last turn: 3  turn=5213203
Num draw() calls last turn: 3  turn=5213203
Num draw() calls last turn: 3  turn=5213204
Num refresh_display() calls last turn: 4  turn=5213204
Num refresh_display() calls last turn: 3  turn=5213205
Num draw() calls last turn: 3  turn=5213205
Num draw() calls last turn: 3  turn=5213206
Num refresh_display() calls last turn: 4  turn=5213206
Num refresh_display() calls last turn: 3  turn=5213207
```

After your change:
```
Num draw() calls last turn: 3  turn=5213249
Num refresh_display() calls last turn: 7  turn=5213250
Num draw() calls last turn: 3  turn=5213250
Num refresh_display() calls last turn: 7  turn=5213251
Num draw() calls last turn: 3  turn=5213251
Num refresh_display() calls last turn: 7  turn=5213252
Num draw() calls last turn: 3  turn=5213252
Num refresh_display() calls last turn: 7  turn=5213253
```

Hopefully this makes more sense to you than it does to me.

**Comment by actual-nh** [comment]
> Hopefully this makes more sense to you than it does to me.

In terms of the line arrangement, it looks like the first group is getting buffered. Depending on your OS, you may be able to get less-buffered output by having it go to STDERR instead of STDOUT; otherwise, some variety of flush/autoflush may be needed.

**Comment by eltank** [comment]
In any case, one way to fix the problem and retain most of your changes is to simply revert the `map.cpp` change. If that was only done in order to allow interrupting auto-drive, then it's not really necessary since auto-drive is already interruptible by pressing `.` and it works well enough.
There may be other solutions of course, but please keep an eye on the number of `refresh_display` calls per turn.

**Comment by eltank** [comment]
> > Hopefully this makes more sense to you than it does to me.
> 
> In terms of the line arrangement, it looks like the first group is getting buffered. Depending on your OS, you may be able to get less-buffered output by having it go to STDERR instead of STDOUT; otherwise, some variety of flush/autoflush may be needed.

I'm pretty sure they're not getting buffered. std::cout << std::endl; forces a flush on my system. The turns are also in chronological order, so I believe that's the true invocation order.

**Comment by actual-nh** [comment]
> I'm pretty sure they're not getting buffered. `std::cout << std::endl;` forces a flush on my system. The turns are also in chronological order, so I believe that's the true invocation order.

Blink. (How refresh_display calls are scheduled in most programs I've looked at, IIRC including CDDA, is to have a flag that gets set when something changes on the intended screen output, and if the flag is set, refresh_display is called and the flag unset.) This may possibly have something (completely obscure) to do with 3+4 = 7, which would make the change per two turns an exact doubling of refresh_display calls.

**Comment by eltank** [comment]
It gets weirder. `handle_key_blocking_activity()` invoked from `map::move_vehicle()` during auto-drive doesn't call `refresh_display()` itself, since it takes the first branch of the `if` and the local variable `refresh` evaluates to false.
I wish I could print a stack trace in `refresh_display` to figure out who's calling it.

**Comment by actual-nh** [comment]
> It gets weirder. `handle_key_blocking_activity()` invoked from `map::move_vehicle()` during auto-drive doesn't call `refresh_display()` itself, since it takes the first branch of the `if` and the local variable `refresh` evaluates to false.
> I wish I could print a stack trace in `refresh_display` to figure out who's calling it.

A profiler or debugger - preferably the former for an initial analysis - should be able to tell that, although I've not worked with either very much at all.

**Comment by eltank** [comment]
I used gdb to figure out how `handle_key_blocking_activity` ends up invoking `refresh_display` indirectly -- it's via `input_context::handle_input` -> `input_manager::get_input_event` -> `catacurses::wnoutrefresh` -> `catacurses::doupdate`

```
#0  refresh_display (caller_file=<optimized out>, caller_line=<optimized out>)
    at src/sdltiles.cpp:558
#1  0x00007ff6d31af04e in input_manager::get_input_event (
    this=this@entry=0x7ff6d3ea7fe0 <inp_mngr>,
    preferred_keyboard_mode=<optimized out>) at src/sdltiles.cpp:3697
#2  0x00007ff6d2caeb5a in input_context::handle_input[abi:cxx11](int) (
    this=this@entry=0x61aa97f0e0, timeout=timeout@entry=0)
    at src/input.cpp:1114
#3  0x00007ff6d2b6c7b4 in handle_key_blocking_activity ()
    at src/do_turn.cpp:569
#4  0x00007ff6d2b70a15 in do_turn () at src/do_turn.cpp:912
#5  0x00007ff6d37ed5e2 in main (argc=<optimized out>, argv=<optimized out>)
    at src/main.cpp:710
```

**Comment by Qrox** [comment]
@eltank I tested with your attached save and still doesn't notice any slowdown.

> revert your change to map.cpp (adding a call to handle_key_blocking_activity in vehicle::move())

Are you by any chance testing with an older commit of this PR? Because in earlier commits `handle_key_blocking_activity` was indeed called in addition to `refresh_display` in `map::move_vehicle`, but the `refresh_display` call was later removed because `handle_key_blocking_activity` already calls it once (via the call to `input_context::handle_input`).

> it's not really necessary since auto-drive is already interruptible by pressing .

`handle_key_blocking_activity` is needed in `map::move_vehicle` because otherwise you cannot interrupt auto-drive during vehicle animation between turns, and any keypress during the animation will be skipped because `inp_mngr.pump_events()` needs to poll window events to handle potential resizing etc.

In any case, `handle_input` is roughly equivalent to `refresh_display` + `pump_events` if there's no keyboard input, so I don't see how that change would make `refresh_display` called more. Profiling also shows that `refresh_display` has the roughly same execution time before and after this change. Counting within `refresh_display` also shows that it's called 3 or 4 times each turn.

**Comment by eltank** [comment]
Hmm, it's possible that I'm using an older version of your code, I'll try re-patching.

In the mean time:
```
    // standards note: getch is sometimes required to call refresh
    // see, e.g., http://linux.die.net/man/3/getch
    // so although it's non-obvious, that refresh() call (and maybe InvalidateRect?) IS supposed to be there

    wrefresh( catacurses::stdscr );
```

I don't really get it but the following also fixes the excessive refresh for me and I didn't see any deleterious effects:
```diff
diff --git a/src/sdltiles.cpp b/src/sdltiles.cpp
index ef1df16..6d1b2ca 100644
--- a/src/sdltiles.cpp
+++ b/src/sdltiles.cpp
@@ -3694,7 +3694,9 @@
     // see, e.g., http://linux.die.net/man/3/getch
     // so although it's non-obvious, that refresh() call (and maybe InvalidateRect?) IS supposed to be there
 
+#if !defined(TILES)
     wrefresh( catacurses::stdscr );
+#endif
 
     if( inputdelay < 0 ) {
         do {
```

**Comment by eltank** [comment]
> handle_key_blocking_activity is needed in map::move_vehicle because otherwise you cannot interrupt auto-drive during vehicle animation between turns

You're not supposed to be able to interrupt autodrive between turns. Once a driving action has been performed (including "pause") you have to wait for `map::move_vehicle` to get called multiple times (depending on the vehicle speed) before you can act again.

**Comment by Qrox** [comment]
> I don't really get it but the following also fixes the excessive refresh for me and I didn't see any deleterious effects:

AFAIK that call is there to display the screen buffer on-screen before polling for input.

> You're not supposed to be able to interrupt autodrive "between turns"

You are right; I didn't express it clear enough. What I mean is that you should be able to press "." during the animation to make it interrupt auto-drive after the animation ends, and calling `handle_key_blocking_activity` allows that.

**Comment by eltank** [comment]
Alright, I re-patched your PR and now I'm getting 3 to 4 `refresh_display` calls per turn (all coming from `cursesport.cpp`) so that performance regression is gone.
Perhaps the regression was never there and I just made a mistake patching last time? Sorry if that's the case.

> You are right; I didn't express it clear enough. What I mean is that you should be able to press "." during the animation to make it interrupt auto-drive after the animation ends, and calling handle_key_blocking_activity allows that.

You could press `.` to interrupt autodrive before, too. If your change makes it more responsive (i.e. react sooner) then that's a win.

File: src/do_turn.cpp [hunk 1]
```diff
@@ -591,9 +589,17 @@ void handle_key_blocking_activity()
...
+    } else {
+        refresh_display();
+        inp_mngr.pump_events();
     }
 }
 
+namespace
+{
+void monmove();
+void overmap_npc_move();
```
Comment (eltank) [inline src/do_turn.cpp:601]: It looks like these forward declarations can be dropped.

```suggestion
```

**Comment by eltank** [comment]
This PR has helped me figure out how to do frame skipping for autodrive: #50803
It's a bit hacky, so I hope you'll be able to take a look and let me know how to make it less hacky.

**Comment by Qrox** [comment]
> I don't really get it but the following also fixes the excessive refresh for me and I didn't see any deleterious effects:

After reading through the code I find that any `wnoutrefresh` call sets `needupdate` on the SDL build, so even if the `wrefresh` call is removed from `input_manager::get_input_event`, `refresh_display` is nevertheless called from `CheckMessages` after the first event poll, which is indistinguishable for a player.

> #50803

This solution looks fine to me. I don't think it's that hacky.

**Pull request merged**